File: src/list/find_max_sublist_sum.ts
// Kadane's
export default function findMaxSublistSum(list: Array<number>) {
  let maximum = -Infinity;
  let current = 0;
  for (const element of list) {
    current += element;
    current = Math.max(current, element);
    maximum = Math.max(current, maximum);
  }
  return maximum;
}

export function findMaxSublistSumBruteforce(list: number[]) {
  let count = 0;
  for (let index = 0; index < list.length; index++) {
    for (let idx = index; idx < list.length; idx++) {
      let localCount = 0;
      for (let i = index; i <= idx; i++) localCount += list[i];
      count = Math.max(count, localCount);
    }
  }
  return count;
}

if (import.meta.vitest) {
  const input = [-10, -20, 10, 20, -10, -50, 10, 20, 40, -20];
  const result = findMaxSublistSum(input);
  it("Should return a number", () => {
    expect(result).toBeTypeOf("number");
    expect(findMaxSublistSumBruteforce(input)).toBeTypeOf("number");
  });
  it("Should not modify the original array", () => {
    expect(input).toEqual([-10, -20, 10, 20, -10, -50, 10, 20, 40, -20]);
    findMaxSublistSumBruteforce(input);
    expect(input).toEqual([-10, -20, 10, 20, -10, -50, 10, 20, 40, -20]);
  });
  it("Should return the correct maximum sublist sum", () => {
    expect(result).toBe(70);
    expect(findMaxSublistSumBruteforce(input)).toBe(70);
    const collection = [-5, -10, 40, 5, -5, 10, -90];
    expect(findMaxSublistSum(collection)).toBe(50);
    expect(findMaxSublistSumBruteforce(collection)).toBe(50);
  });
}

File: src/list/find_max_sublist_sum.py
import sys

# Kadane's
def find_max_sublist_sum(arr: list[int]):
  maximum = ~sys.maxsize
  current = 0
  for element in arr:
    current += element
    current = max(current, element)
    maximum = max(current, maximum)
  return maximum

def find_max_sublist_sum_bruteforce(list: list[int]):
  count = 0
  for index in range(0, len(list)):
    for idx in range(index, len(list)):
      local_count = 0
      for i in range(index, idx): 
        local_count += list[i]
      count = max(count, local_count)
  return count

input = [-10, -20, 10, 20, -10, -50, 10, 20, 40, -20]
assert find_max_sublist_sum(input) == 70
assert find_max_sublist_sum_bruteforce(input) == 70
input = [-5, -10, 40, 5, -5, 10, -90]
assert find_max_sublist_sum(input) == 50
assert find_max_sublist_sum_bruteforce(input) == 50
